examples/pruning/lightgbm_with_custom_cv_function.py

```diff
@@ -0,0 +1,404 @@
+times with unaffected accuracy.
+
+The example emphasizes reproducibility, giving the user precise control over all seed
+values that are being used to let her arrive at the same results over repeated studies.
+
+You can run this example using python 3 interpreter, e.g. as follows:
+    $ python lightgbm_with_custom_cv_function.py
+"""
+
+# Import packages
+
+from datetime import datetime
+from pprint import pprint
+
+from lightgbm import LGBMClassifier
+from lightgbm import LGBMRegressor
+import numpy as np
+import pandas as pd
+from sklearn import datasets
+from sklearn.model_selection import KFold
+from sklearn.model_selection import train_test_split
+
+import optuna
+from optuna.pruners import MedianPruner
+from optuna.samplers import TPESampler
+
+
+# Custom functions
+
+
+def custom_cv_fun(
+    lgbm_params,
+    X,
+    y,
+    objective="binary",
+    eval_metric="auc",
+    eval_name="valid",
+    n_estimators=100,
+    early_stopping_rounds=10,
+    nfold=5,
+    random_state=123,
+    callbacks=[],
+    verbose=False,
+):
+
+    # create placeholders for results
+    fold_best_iterations = []
+    fold_best_scores = []
+
+    # get feature names
+    feature_names = list(X.columns)
+
+    # split data into k folds
+    kf = KFold(n_splits=nfold, shuffle=True, random_state=random_state)
+    splits = kf.split(X, y)
+
+    # iterate over folds
+    for n, (train_index, valid_index) in enumerate(splits):
+
+        # subset train and valid (out-of-fold) parts of the fold
+        X_train, X_valid = X.iloc[train_index], X.iloc[valid_index]
+        y_train, y_valid = y.iloc[train_index], y.iloc[valid_index]
+
+        # select estimator for the current objective
+        lgbm_cv_fun = None
+        if objective in ["binary", "classification"]:
+            lgbm_cv_fun = LGBMClassifier
+        elif objective == "regression":
+            lgbm_cv_fun = LGBMRegressor
+
+        model = lgbm_cv_fun(n_estimators=n_estimators, random_state=random_state)
+
+        # pass hyperparameters dict to the estimator
+        model.set_params(**lgbm_params)
+
+        # train the model
+        model.fit(
+            X_train,
+            y_train.values.ravel(),
+            eval_set=(X_valid, y_valid.values.ravel()),
+            eval_metric=[eval_metric],  # note: a list required
+            eval_names=[eval_name],  # note: a list required
+            early_stopping_rounds=early_stopping_rounds,
+            verbose=verbose,
+            feature_name=feature_names,
+            callbacks=callbacks,
+        )
+
+        # collect current fold data
+        fold_best_iterations.append(model.best_iteration_)
+        fold_best_scores.append(model.best_score_[eval_name])
+
+    # average folds iterations numbers
+    folds_data = {}
+    folds_data["best_iterations_mean"] = -1
+    if fold_best_iterations[0] is not None:
+        folds_data["best_iterations_mean"] = int(np.mean(fold_best_iterations))
```

Comment:
[Question] Could you tell me when the `model.best_iteration_` returns `None`?

Reply:
Apologies - here the comment would be probably in order. This occurs when the user switches off early stopping (by setting `USE_STOPPING=False`, and then the `fold_best_iterations` list is populated with `None` elements (as returned by `model.best_iteration_` in such cases).

Reply:
Thank you for adding the comment to L107 :+1: